File: code/goldbach_tuple_hypothesis.py
import math

def sieve_of_eratosthenes(limit):
    """Generate primes up to 'limit' using Sieve of Eratosthenes."""
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False
    for num in range(2, int(math.sqrt(limit)) + 1):
        if sieve[num]:
            sieve[num*num : limit+1 : num] = [False] * len(sieve[num*num : limit+1 : num])
    return [i for i, is_prime in enumerate(sieve) if is_prime]

def test_binary_interval_hypothesis(k_max):
    """Test the hypothesis for all binary intervals up to k=k_max."""
    for k in range(2, k_max + 1):
        lower = 2 ** k
        upper = 2 ** (k + 1)
        primes = sieve_of_eratosthenes(upper)
        primes_up_to_half = [p for p in primes if p <= 2 ** (k - 1)]
        
        failed_evens = []
        for N in range(lower, upper + 1, 2):  # Even numbers only
            found = False
            for p in primes_up_to_half:
                q = N - p
                if q in primes:  # Check if q is prime (using set for O(1) lookup)
                    found = True
                    break
            if not found:
                failed_evens.append(N)
        
        print(f"\nResults for [2^{k}, 2^{k+1}):")
        #print(f"Primes ≤ 2^{k-1}: {primes_up_to_half}")
        if not failed_evens:
            print(f"All evens in [2^{k}, 2^{k+1}] satisfy the hypothesis!")
        else:
            print(f"Counterexamples found: {failed_evens}")

# Test up to k=6 (i.e., N ∈ [2^6, 2^7] = [64, 128])
test_binary_interval_hypothesis(k_max=18)

print("End of calculations.")
print("")

File: code/goldbach_tuple_hypothesis.cpp
#include <iostream>
#include <vector>
#include <bitset>
#include <cmath>
#include <algorithm>
#include <chrono>
#include <cstdint>

void test_hypothesis_original_behavior(int k_max) {
    for (int k = 2; k <= k_max; ++k) {
        const uint64_t lower = 1ULL << k;
        const uint64_t upper = (1ULL << (k + 1)) - 1;
        
        // Generate primes up to upper
        std::vector<bool> sieve(upper + 1, true);
        sieve[0] = sieve[1] = false;
        for (uint64_t p = 2; p * p <= upper; ++p) {
            if (sieve[p]) {
                for (uint64_t multiple = p * p; multiple <= upper; multiple += p) {
                    sieve[multiple] = false;
                }
            }
        }

        // Collect primes <= 2^(k-1)
        const uint64_t half_limit = 1ULL << (k - 1);
        std::vector<uint64_t> primes_half;
        for (uint64_t p = 2; p <= half_limit; ++p) {
            if (sieve[p]) primes_half.push_back(p);
        }

        // Test even numbers
        std::vector<uint64_t> failed_evens;
        for (uint64_t N = lower; N <= upper; N += 2) {
            bool found = false;
            for (uint64_t p : primes_half) {
                if (p > N) break;  // Safety check
                const uint64_t q = N - p;
                if (q >= 2 && q <= upper && sieve[q]) {
                    found = true;
                    break;
                }
            }
            if (!found) failed_evens.push_back(N);
        }

        // Original-style output
        std::cout << "\nInterval [2^" << k << ", 2^" << (k + 1) << "):\n";
        std::cout << "Primes <= 2^" << (k - 1) << ": " << primes_half.size() << " primes\n";
        if (failed_evens.empty()) {
            std::cout << "All evens satisfy the hypothesis!\n";
        } else {
            std::cout << "Counterexamples: ";
            for (auto N : failed_evens) std::cout << N << " ";
            std::cout << "\n";
        }
    }
}

int main() {
    int k_max;
    std::cout << "Enter maximum k (e.g., 20): ";
    std::cin >> k_max;

    auto start = std::chrono::high_resolution_clock::now();
    test_hypothesis_original_behavior(k_max);
    auto end = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double> elapsed = end - start;
    std::cout << "\nTime taken: " << elapsed.count() << " seconds\n";
    return 0;
}